frame/scheduler/src/lib.rs

```diff
@@ -105,31 +108,145 @@ decl_storage! {
+		/// # </weight>
+		#[weight = 25_000_000 + T::DbWeight::get().reads_writes(1, 1)]
+		fn schedule(origin,
+			when: T::BlockNumber,
+			maybe_periodic: Option<schedule::Period<T::BlockNumber>>,
+			priority: schedule::Priority,
+			call: Box<<T as Trait>::Call>,
+		) {
+			ensure_root(origin)?;
+			let _ = Self::do_schedule(when, maybe_periodic, priority, *call);
+		}
+
+		/// Cancel an anonymously scheduled task.
+		///
+		/// # <weight>
+		/// - S = Number of already scheduled calls
+		/// - Base Weight: 22.15 + 2.869 * S µs
+		/// - DB Weight:
+		///     - Read: Agenda
+		///     - Write: Agenda, Lookup
+		/// - Will use base weight of 100 which should be good for up to 30 scheduled calls
+		/// # </weight>
+		#[weight = 100_000_000 + T::DbWeight::get().reads_writes(1, 2)]
+		fn cancel(origin, when: T::BlockNumber, index: u32) {
+			ensure_root(origin)?;
+			Self::do_cancel((when, index))?;
+		}
+
+		/// Schedule a named task.
+		///
+		/// # <weight>
+		/// - S = Number of already scheduled calls
+		/// - Base Weight: 29.6 + .159 * S µs
+		/// - DB Weight:
+		///     - Read: Agenda, Lookup
+		///     - Write: Agenda, Lookup
+		/// - Will use base weight of 35 which should be good for more than 30 scheduled calls
+		/// # </weight>
+		#[weight = 35_000_000 + T::DbWeight::get().reads_writes(2, 2)]
+		fn schedule_named(origin,
+			id: Vec<u8>,
+			when: T::BlockNumber,
+			maybe_periodic: Option<schedule::Period<T::BlockNumber>>,
+			priority: schedule::Priority,
+			call: Box<<T as Trait>::Call>,
+		) {
+			ensure_root(origin)?;
+			Self::do_schedule_named(id, when, maybe_periodic, priority, *call)?;
+		}
+
+		/// Cancel a named scheduled task.
+		///
+		/// # <weight>
+		/// - S = Number of already scheduled calls
+		/// - Base Weight: 24.91 + 2.907 * S µs
+		/// - DB Weight:
+		///     - Read: Agenda, Lookup
+		///     - Write: Agenda, Lookup
+		/// - Will use base weight of 100 which should be good for up to 30 scheduled calls
+		/// # </weight>
+		#[weight = 100_000_000 + T::DbWeight::get().reads_writes(2, 2)]
+		fn cancel_named(origin, id: Vec<u8>) {
+			ensure_root(origin)?;
+			Self::do_cancel_named(id)?;
+		}
+
+		/// Execute the scheduled calls
+		///
+		/// # <weight>
+		/// - S = Number of already scheduled calls
+		/// - N = Named scheduled calls
+		/// - P = Periodic Calls
+		/// - Base Weight: 9.243 + 23.45 * S µs
+		/// - DB Weight:
+		///     - Read: Agenda + Lookup * N + Agenda(Future) * P
+		///     - Write: Agenda + Lookup * N  + Agenda(future) * P
+		/// # </weight>
 		fn on_initialize(now: T::BlockNumber) -> Weight {
 			let limit = T::MaximumWeight::get();
 			let mut queued = Agenda::<T>::take(now).into_iter()
 				.enumerate()
 				.filter_map(|(index, s)| s.map(|inner| (index as u32, inner)))
 				.collect::<Vec<_>>();
 			queued.sort_by_key(|(_, s)| s.priority);
-			let mut result = 0;
+			let base_weight: Weight = T::DbWeight::get().reads_writes(1, 1) // Agenda
```

Comment:
In worst case we do `Agenda::take(now)` and `Agenda::append(now + one)`
```suggestion
			let base_weight: Weight = T::DbWeight::get().reads_writes(1, 2) // Agenda
```